gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -215,6 +228,25 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
         in.beginObject();
         while (in.hasNext()) {
           String name = in.nextName();
+          if (this.defaultTyping && TYPE_FIELD_NAME.equals(name)) {
+            String className = in.nextString();
+            if (PROPERTIES_FIELD_NAME.equals(in.nextName())) {
+              try {
+                Class<?> clazz = Class.forName(className, false, this.getClass().getClassLoader());
+                if (!instance.getClass().equals(clazz) && instance.getClass().isAssignableFrom(clazz)) {
```

Comment:
Maybe it would be saner to throw an exception (`JsonParseException`) if `instance.getClass().isAssignableFrom(clazz)` is not the case? Currently it would just deserialize as if the type name did not exist which might make noticing this issue more difficult.